linked_list/linked_list.py

```diff
@@ -9,67 +9,167 @@ def __init__(self, value, next_node = None):
     def add_first(self, value):
-        pass
+        
+        #create a new node
+        new_node = Node(value) 
+        #set the new node to point to the current head
+        # new_node.next = self.head
+        # #Assign the new node to the head variable
+        # self.head = new_node
+
+        #How could this be performed in one line?
+        self.head = Node(value, next_node=self.head)
 
     # method to find if the linked list contains a node with specified value
     # returns true if found, false otherwise
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: o(n)/linear complexity
+    # Space Complexity: o(n)/linear complexity
     def search(self, value):
-        pass
+
+        if self.head == None:
+            return False
+
+        current_node = self.head
+        while current_node.next != None:
+            if current_node.value == value:
+                return True
+            current_node = current_node.next
+        
+        if current_node.value == value:
+            return True
+        return False
 
     # method that returns the length of the singly linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: o(n)/linear complexity
+    # Space Complexity: o(n)/linear complexity
     def length(self):
-        pass
+        count = 0
+        current_node = self.head
+
+        while current_node != None:
+            count = count + 1
+            current_node = current_node.next
+        
+        return count
 
     # method that returns the value at a given index in the linked list
     # index count starts at 0
     # returns None if there are fewer nodes in the linked list than the index value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: o(n)/linear complexity
+    # Space Complexity: o(n)/linear complexity
     def get_at_index(self, index):
-        pass
+        current_node = self.head
+        count = 0
+
+        while current_node != None:
+            if count == index:
+                return current_node.value
+
+            count += 1
+            current_node = current_node.next
+        
+        return None
+
 
     # method that returns the value of the last node in the linked list
     # returns None if the linked list is empty
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: o(n)/linear complexity
+    # Space Complexity: o(n)/linear complexity
     def get_last(self):
-        pass
+        
+        #make sure list is not empty
+        if self.head == None:
+            return None
+
+        current_node = self.head
+        while current_node.next != None:
+            current_node = current_node.next
+        
+        return current_node.value
 
     # method that inserts a given value as a new last node in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def add_last(self, value):

```

Comment:
👍 , However since you're only creating one new node the space complexity is O(1)